### DISCUSSION TITLE
Obitools NGS filter

### DISCUSSION THREAD

**Discussion by skreling** [sr320/course-fish546-2021#109] [Q&A] [answered]
Having trouble getting the ngsfilter to work for obitools. I created my own based on what the paper said the primers [were](https://scholar.google.com/scholar_lookup?journal=Mol.+Ecol.+Resour&title=Taxon-specific+PCR+for+DNA+barcoding+arthropod+prey+in+bat+faeces&author=MRK+Zeale&author=RK+Butlin&author=GLA+Barker&author=D+Lees&author=G+Jones&volume=11&publication_year=2011&pages=236-244&pmid=21429129&doi=10.1111/j.1755-0998.2010.02920.x&). However, I don't know what specific Illumina tags are on the samples, so I just didn't include a 'tags' column in the ngs filter. Also the only info I can find on creating a filter is [here](https://pythonhosted.org/OBITools/scripts/ngsfilter.html) and I think I followed it's formatting guidelines. 

[Here](https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Scripts/01_Bat_ObiTools_2.ipynb) is the Jupyter notebook I'm working from and [here](https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Data/Bat_Data/ngs_filter_bat.txt) is the NGS filter and here is the sample ([forward](https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Data/Bat_Data/2_S2_L001_R1_001.fastq.gz) and [reverse](https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Data/Bat_Data/2_S2_L001_R2_001.fastq.gz)) that I was trying to process (note these are compressed).

And here is just a quick screen shot of the error, it's also in that Jupyter notebook. I attempted to proceed with just the 'good_sequences' obi view, to dereplicate, but it wouldn't proceed without the 'identified_sequencess' created.
<img width="887" alt="Screen Shot 2021-02-28 at 4 29 08 PM" src="https://user-images.githubusercontent.com/76884520/109439581-cbe12880-79e3-11eb-9fb2-aec59b743d0e.png">

And [here](https://git.metabarcoding.org/obitools/obitools3/wikis/Wolf-tutorial-with-the-OBITools3) is the general workflow.

**Comment by skreling** [comment]
Also found out that the --no-tags argument should be there, but still getting an error message, but now it just looks like my arguments are incorrect or I'm no laying it out correctly?

<img width="865" alt="Screen Shot 2021-02-28 at 4 59 55 PM" src="https://user-images.githubusercontent.com/76884520/109440244-81ad7680-79e6-11eb-972c-a7c06acd0ebf.png">

**Reply by sr320** [reply]
looks like you might have inadvertently removed the `-t` flag? that is required.

**Reply by skreling** [reply]
Yes, I did also realize that so I changed i to `!obi ngsfilter -t bat/ngsfile -u bat/unidentified_sequences --no-tags bat/good_sequences bat/identified_sequences` but then get the same KeyError: 'Cannot access to column reverse_primer in view ngsfile'

**Reply by sr320** [reply]
is `bat/ngsfile` in your repo - can you provide the url

**Reply by skreling** [reply]
It wasn't because it's essentially the ngs_filter_bat.txt file that I linked to in the original post, but I've pushed some of the obi [files](https://github.com/fish546-2021/Sam-Metabarcoding/tree/main/Data/Bat_Data/bat.obidms/VIEWS) up. Obitools works a bit weirdly and creates its own file system with everything in .obiview extensions. Not actually sure what its doing behind the scenes.

**Reply by sr320** [reply]
I think it is looking for a `reverse_primer` column.. and while very very hard to decipher I see no sign of one. https://d.pr/i/uowUhu (though do see `forward_primer`)

**Reply by skreling** [reply]
Ok, so based on that information I think you're right and I had a bit of a search through the files, I've pushed a few more of the relevant [ones](https://github.com/fish546-2021/Sam-Metabarcoding/tree/main/Data/Bat_Data/bat.obidms) up. 

It does appear that it isn't populating the reverse_primer.obicol. There are two entries in there, but from a previous time that I ran this with different data, so it's not relevant to this per say, but shouldn't be hindering anything either. 

I also saw in the Jupyter Notebook that it looks like the progress bar only gets to 67% or so under the `!obi import --fastq-input 2_S2_L001_R2_001.fastq bat/reads2` but I'm not sure why it wouldn't be populating the reverse_primer column or how to go about fixing it. The manual for obitools has not been particularly helpful for errors

**Comment by kubu4** [comment, marked as answer]
I haven't worked my way through this yet, so I'm just going to do a "stream-of-consciousness" type thing where I post things I find as  I go along. 

The first thing I noticed is the formatting of the `ngs_filter_bast.txt` file seems like it's screwed up:

![image](https://user-images.githubusercontent.com/4514104/109520272-e5748580-7a60-11eb-9d0b-63d722ecfd82.png)


I've circled two problems with this file:

1. There's a strange, non-text character embedded at the beginning of each sequence. 

2. The file doesn't end with a newline (usually Unix-based text files end with a newline). This is evidenced by the fact that the command new command prompt begins after printing the contents of the file, instead of showing the command prompt on it's own, new line.

**Reply by kubu4** [reply]
Actually, that's as far as I'll go for now. Fix that issue and things will probably work better. Try it out and report back. I'll work through more issues if that doesn't fix things.

Also, make sure you're only editing/creating text files using Atom, if possible (or, another "proper" text editor). A character of that nature appearing in a text file is usually due to some other type of editor (e.g. Microsoft Word/Excel) being used.

**Reply by skreling** [reply]
I went back and just completely remade the [ngs filter](https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Data/Bat_Data/bat_ngs_atom.txt) in atom and I'm still getting exactly the same errors. I've pushed the rest of the [obitools file structure ](https://github.com/fish546-2021/Sam-Metabarcoding/tree/main/Data/Bat_Data/bat.obidms)up and it seems to be populating the forward primer, forward tags, reverse tags, but not reverse primers folder. (as an aside the number of files in each of those folders and what number they end with is just the amount of times I've run the same common and it's created an additional file). Note that the Jupiter notebook here isn't updated with the change, but I did change the file and received the exact same error and it still only imported 67% it looks  like when I loaded the filter.

**Reply by kubu4** [reply]
Alrighty, got past the problem area!

I've pushed an updated (i.e. working!) version of your `bat_ngs_atom.txt` file.

As it turns out, the `F` in the `extra_information` column is _required_!

I also removed the second row, as that is actually the same sample; it's just the second set of reads. Feel free to add it back in, if you want (but don't forget to add an `F` in the `extra_information` column!).

With those changes, I was able to successfully run:

`obi ngsfilter -t bat/ngsfile -u bat/unidentified_sequences bat/good_sequences bat/identified_sequences`

I have _not_ tested beyond this. 

Anyway, give it a whirl and let us know if it works.

**Comment by skreling** [comment]
That worked thank you! They really should have better documentation, I'm not sure I ever would have figured out that that 'F' was mandatory.

New problem (sorry) : Can't get the dereplication to work properly. Updated notebook [here](https://github.com/fish546-2021/Sam-Metabarcoding/blob/main/Scripts/01_Bat_ObiTools_2.ipynb).
<img width="1107" alt="Screen Shot 2021-03-01 at 11 51 17 AM" src="https://user-images.githubusercontent.com/76884520/109550668-70b34280-7a84-11eb-8d78-467df78ae674.png">

**Reply by kubu4** [reply]
Weird. I ran that command without issue:

![image](https://user-images.githubusercontent.com/4514104/109551079-f3d49880-7a84-11eb-8339-839c8384b78f.png)


Personally, I'd recommend deleting your `bat.obidms` directory and start fresh.

**Reply by skreling** [reply]
I just deleted as you suggested and rebuilt it, but I'm still getting the exact same error?

**Reply by sr320** [reply]
really random ideas - @skreling try running outside of Jupyter..

Could it be python version related?

**Reply by kubu4** [reply]
I started fresh and am getting the same error as you. Will look into it and report back if/when I find something.

**Reply by kubu4** [reply]
Got it! Run the command _without_ the `-m sample` argument!

![image](https://user-images.githubusercontent.com/4514104/109660345-96d8f100-7b1d-11eb-9088-671fdfc82d6b.png)

My rationale (although I haven't tested) is that your sample only has a single attribute, so no need to use that argument.

**Reply by skreling** [reply]
Whew this has been a rollercoaster. I ended up uninstalling obitools and reinstalling python3/reorgainzing file structures because things were in the wrong place and that was making things more complicated (though it took me a full 18hr to figure out where to put them correctly and get obitools reinstalled) but finally got it back and tried the above suggestions and nothing has worked!

If you run without the `-m sample` it brings up the manual for the command, so that appears to be a mandatory argument. 

Also tried running outside of Jupyter and it appears to be faster, but still getting the same error.

**Reply by kubu4** [reply]
> so that appears to be a mandatory argument.

Definitely not mandatory - see my screencap in the above post. No `-m` argument and it ran without issue.

Installing/uninstalling Python3? Oof, that sounds horribly complicated/problematic. Why aren't you using a conda environment any more?

**Reply by kubu4** [reply]
Also, is it safe to assume that you're getting the same error message as before?

**Reply by skreling** [reply]
It wasn't great, but I improperly installed python 3 to begin with so it was nice to fix that issue. I don't really have an answer as to why I chose not to use a conda venv, now I'm just using a python venv and everything is finally working well until this point

Yes same error message as before with the keyError: b'MERGED_sample'

**Reply by kubu4** [reply]
Hmmm, interesting (and frustrating - sorry!).

Have you tried running through everything fresh (i.e. deleting your `bat.obidms` folderr)?

**Reply by skreling** [reply]
yes sadly that didn't work either :(

**Reply by kubu4** [reply]
Wanna give it a go by installing ObiTools3 in a conda environment? That way we'd know that we're both working on the same exact page, since I'm working in a conda environment.

**Reply by skreling** [reply]
Still the same error even in conda environment
<img width="567" alt="Screen Shot 2021-03-02 at 10 13 39 AM" src="https://user-images.githubusercontent.com/76884520/109694775-08299b80-7b40-11eb-8df3-ad8199e4b2b5.png">

**Reply by kubu4** [reply]
You need to run it _without_ the `-m sample` argument.

**Reply by skreling** [reply]
That worked this time! Thank you sorry this was such a hassle!
